---
title: "Bachelors Project"
format: html
editor: visual
---

# Investigating Cross-Reactivity Between Autoimmunogenic Epitopes and Pathogen Proteomes

## Summary

The project aims to investigate cross-reactivity between known autoimmunogenic epitopes, sourced from the IEDB database, and proteomes of pathogenic bacteria and viruses. To manage the complexity, I will cluster bacterial and viral proteomes using MMseqs2, selecting representative proteins from each group for analysis.

I will then, from biological knowledge, weigh the different amino acids positions. The usual culprit in cross reactivity is the MHCII which usually present 15mers. We know the usual hydrophobic clefts position in MHCII and the positions that are in contact with the TCR. From this we can weigh importance of specific positions. We will make a model that predicts the cross reactivity for a given protein, and we will test it on know autoimmunogenic proteins from pathogens.

## Method

Using IEDB i will find the epitopes that are found to be associated with all autoimmune diseases in humans.

The following filters where applied: "Epitope Structure", "Include Positive Assays", "MHC Restriction Type: Class II", "Host: Homo Sapiens (Human)", "Disease Data: Autoimmune Disease", "No T cell assays" and "No B cell assays".

These filters will ensure that I get only Linear MHC class II epitopes, that is found only with MHC assays, meaning that i will be able to find to MHC allele that was used. This will allow me to find the binding core, using netpanMHCII. and then find the anchor positions on the peptides. With the found anchor positions we can now weigh the conservation of this position more, allowing for more accurate cross-reactivity check.

This Returned \~? epitopes for analyzation.

### libraries

```{r}
#| message: false
#| warning: false

library("tidyverse")
library("Biostrings")
library("ggseqlogo")
library("biomaRt")
library("pdftools")
library("magick")
```

### Data import

```{r}
#| message: false
#| warning: false

source("./functions.R")

# Reading in headers
headers <- read_csv("./Data/autoimmune_data.csv", n_max = 2, col_names = FALSE)

# Combine the first two rows into a single header
new_col_names <- paste(headers[1,], headers[2,], sep = "_")
new_col_names <- gsub("_NA$", "", new_col_names)

autoimmune_data <- read_csv("./Data/autoimmune_data.csv", skip = 2, col_names = new_col_names)

# From class will be replaced with bigger bacteria proteome dataset
pathogen_fasta <- readAAStringSet("./Data/IEDB_infectius_peptides.fsa",)

# Clustered bacteria proteome dataset using mmseq2
pathogen_clusters <- read_tsv("./Data/infectius_cluster.tsv", col_names = c("cluster_ID", "member_ID"))

# From NetMHCIIpan containing information about binding core
DRB5_0101_NetMHCIIpan <- read_csv("./Data/DRB5_0101_NetMHCIIpan.csv", skip = 1, col_names = TRUE)
DRB1_1501_NetMHCIIpan <- read_csv("./Data/DRB1_1501_NetMHCIIpan.csv", skip = 1, col_names = TRUE)
DRB1_0401_NetMHCIIpan <- read_csv("./Data/DRB1_0401_NetMHCIIpan.csv", skip = 1, col_names = TRUE)

# Experimentally verifiyed pathogenic peptides causing autoimmune diseases
exp_verified_patho_peps <- read_csv("./Data/Exp_verified_pathogenic_peptides.txt")
```

### Wrangling IEDB tables for netMHCIIpan

```{r}
known_diseases <- c('multiple sclerosis',"type 1 diabetes mellitus", "rheumatoid arthritis", "myasthenia gravis","systemic lupus erythematosus", "Graves' disease, Graves disease", "autoimmune hemolytic anemia", "psoriasis", "autoimmune uveitis", "autoimmune thyroiditis", "vitiligo", "alopecia areata", "Vogt-Koyanagi-Harada disease", "Goodpasture syndrome", "Behcet's disease", 'systemic lupus erythematosus', 'neuromyelitis optica', 'primary biliary cholangitis', 'autoimmune hepatitis', 'autoimmune vasculitis', 'pemphigus',"Sjogren's syndrome")


# Tidying data and filtering with relevant filters
autoimmune_data <- autoimmune_data |>
  filter(`Epitope_Source Organism` == "Homo sapiens") |> # Remove this and add filter in IEDB
  filter(is.na(`Epitope_Modified residues`)) |>
  dplyr::rename("Assay_ID" = "Assay ID_IEDB IRI",
                "Epitope_ID" = "Epitope_IEDB IRI",
                "Epitope" = "Epitope_Name",
                "Epitope_Protein" = "Epitope_Molecule Parent",
                "Epitope_Protein_annot" = "Epitope_Molecule Parent IRI",
                "Disease" = "1st in vivo Process_Disease",
                "MHC_Allele" = "MHC Restriction_Name") |>
  mutate(Epitope = str_trim(str_replace(Epitope, "\\+.*", "")),
         Assay_ID = as.integer(str_trim(str_extract(Assay_ID, "\\d+$"))),
         Epitope_ID = as.integer(str_trim(str_extract(Epitope_ID, "\\d+$"))),
         Epitope_Protein_annot = str_trim(str_extract(Epitope_Protein_annot, "(?<=/)[^/]+$"))) |>
  dplyr::select(Assay_ID, Epitope_ID, Epitope, Epitope_Protein, Epitope_Protein_annot, Disease, MHC_Allele) |>
  filter(str_length(Epitope) > 12 & str_length(Epitope) < 26) |>
  filter(!is.na(MHC_Allele) & !is.na(Epitope)) |> filter(str_detect(Epitope, "^[ACDEFGHIKLMNPQRSTVWY]+$")) |>
  filter(Disease %in% known_diseases) |>
  unique()

autoimmune_data |>
  group_by(MHC_Allele) |>
  summarise(amount = n()) |>
  arrange(desc(amount))
```

### Writing fastas:

```{r}
# Write fasta file for netMHCIIpan for the chosen allele of interest
write_fasta("HLA-DRB5\\*01:01", autoimmune_data, "Assay_ID", "Epitope", "./Data/DRB5_0101.fasta")
write_fasta("HLA-DRB1\\*15:01", autoimmune_data, "Assay_ID", "Epitope", "./Data/DRB1_1501.fasta")
write_fasta("HLA-DRB1\\*04:01", autoimmune_data, "Assay_ID", "Epitope", "./Data/DRB1_0401.fasta")
```

### NetMHCIIpan output wrangling and decision of anchorpoints:

based on www.syfpeithi.de we found that typical anchorpoints for DRB5_0101 are

-   P1: F, Y, L, M

-   P4: Q, V, I, M

-   P9: R, K

    We can therefore decide to check cross-reactivity based on this or look at our peptides that was binder found through NetMHCIIpan by a logoplot.

```{r}

DRB1_0401_motifs <- MHC_motif_finder("DRB1_0401")
DRB1_1501_motifs <- MHC_motif_finder("DRB1_1501")
DRB5_0101_motifs <- MHC_motif_finder("DRB5_0101")
```

### Cluster wrangeling using mmseqs2:

using mmseqs and the IEDB_infectius_peptides.fsa, we cluster based on sequence similarity using a minimum sequence identity of 70% and a 80% alignment coverage

```{bash}
#| eval: false

mmseqs createdb unique_IEDB_infectius_peptides.fsa infectius_peptides_DB

mmseqs cluster infectius_peptides_DB infectius_cluster tmp

mmseqs createtsv infectius_peptides_DB infectius_peptides_DB infectius_cluster infectius_cluster.tsv
```

The discovery from this .fsa file was, that the sequences that had the same accession name was clustered together (which makes sense). Therefore we will simply take the original fasta, group them and pick 1 representative. This will reduce the amount of peptides we need to compare to our known autoimmune epitopes, allowing for faster computing.

### Tidying pathogen proteome data

```{r}

# Making the pathogen proteome fastafile to a dataframe
pathogen_seqs <- data.frame(
    header = names(pathogen_fasta),
    sequence = as.character(pathogen_fasta)
)

# Filtering pathogen proteome based on the found anchors
pathogen_seqs_DRB5_0101 <- pathogen_seqs |>
  filter(str_detect(sequence, "[FYILVW].{2}[ALIVEQ].{1}[ASGPNI].{2}[KR]"))

pathogen_seqs_DRB1_1501 <- pathogen_seqs |>
  filter(str_detect(sequence, "[ILVM].{2}[YFW].{1}[SNGDA].{2}[LVAI]"))

pathogen_seqs_DRB1_0401 <- pathogen_seqs |>
  filter(str_detect(sequence, "[YFIW].{2}[DEAQM].{1}[TNSDV].{2}[ASGQ]"))
```

### Checking for cross-reactivity:

```{r}
percent_identity <- 50

# Running cross reactivity test
DRB5_0101_results <- cross_reactive_test(50, autoimmune_data, pathogen_seqs_DRB5_0101, 
                      "HLA-DRB5\\*01:01")
DRB1_1501_results <- cross_reactive_test(50, autoimmune_data, pathogen_seqs_DRB1_1501, 
                      "HLA-DRB1\\*15:01")
DRB1_0401_results <- cross_reactive_test(50, autoimmune_data, pathogen_seqs_DRB1_0401, 
                      "HLA-DRB1\\*04:01")

DRB5_0101_results
DRB1_1501_results
DRB1_0401_results
```

### Converting gene name for comparison

```{r}
#| eval: false

# Function to get organism from UniProt ID using biomaRt
get_organism_from_uniprot_biomart <- function(uniprot_id) {
  # Use ENSEMBL's mart and dataset for human genes
  ensembl <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl")
  
  # Query the UniProt ID to get organism information
  organism_info <- getBM(attributes = c("uniprot_swissprot", "organism_name"),
                         filters = "uniprot_swissprot",
                         values = uniprot_id,
                         mart = ensembl)
  
  # Return organism name
  return(organism_info$organism_name)
}

# Example usage
uniprot_id <- "P03211"  # Replace with your UniProt ID
organism <- get_organism_from_uniprot_biomart(uniprot_id)


autoimmune_data |>
  group_by(Disease) |>
  summarise(n())
```

```{r}
```
